### ISSUE TITLE
Multiple save profiles

### ISSUE THREAD

**Issue by tony971** [mgba-emu/mgba#106]
If you run multiple windows of the same game, they will use the same .sav file. This could be problematic for games like Four Swords Adventures or Pokemon, which are designed to be played with other cartridges of the same title.

Edit: My last mockup is much better.

**Comment by tony971** [comment]
You'd probably need to force an emulation reset after switching profiles. That'll need a confirmation/warning.

**Comment by tony971** [comment]
https://moqups.com/tony971/QrOwORV0

**Comment by tony971** [comment]
The only thing that's giving me any turmoil is whether or not this should be select-able before loading a ROM. Because if you intend to load into a different profile, you'd need that at the start of emulation. But making it a universal setting isn't a good idea either. Users could always reset after changing slots, I suppose. But that shouldn't be mandated because some users might switch slots mid-play just to keep a copy of a save before some big event.

**Comment by endrift** [comment]
I'm really not sure how users are expected to understand this functionality, and especially how to handle the file backing for it. I can try to figure something out, but I'm not entirely sure what the best approach is. At the moment, I'm just making a copy of the ROM and the save file, then loading both.

**Comment by karasuhebi** [comment]
I don't think this is intuitive enough for users. Needs to be made better. But I do support the general idea of it.

**Comment by tony971** [comment]
I just don't know how to make it more intuitive :( each profile would act like it's own GBA. The other idea I had was naming it "Select Active GBA" with the options as "GameBoy Advance [x]" but changed my mind when I thought users would get confused about when they were supposed to switch between them for multiplayer.

**Comment by tony971** [comment]
I know that VBA-M just makes it so Player 2 always uses .sav2 files but that's even harder to use correctly than this setting. Because unless you always play a game as player 2, you'll be using the .sav slot all the way until you try to use it in multiplayer.

**Comment by karasuhebi** [comment]
Maybe make it so mGBA doesn't allow two ROMs to use the same .sav file (lock it maybe?) and just warn the user when they try to open a second copy of the game that they need to have a .sav2 for it.

**Comment by tony971** [comment]
That still has the same problem as VBA-M, where you can't use the .sav2 file until you're in multiplayer (unless you edit file extensions in your ROM library).

**Comment by tony971** [comment]
You could combine the two ideas. Use my menu item to use whichever save file you want in single player but then default it to Profile 2 if a multiplayer window is opened with the same ROM as Player 1.

**Comment by tony971** [comment]
https://moqups.com/tony971/QrOwORV0

I've updated the  mockup to be a little more straightforward. If .sav[#of windows running the same ROM) is defaulted, then this could be intuitive enough for a solution. I would just have  an OSD saying  "rom.sav[x] loaded" whenever this automatic default happens.

**Comment by tony971** [comment]
Wait. I've got it! You label it as "Active Cartridge" with options of "Cartridge 1 (.sav)" "Cartridge 2 (.sav2)" "..." Most users should understand that.

**Comment by karasuhebi** [comment]
I agree that a .sav2 may not be the best solution, but I just think that your solution is making it more complicated than it needs to be. We have to keep it simple for end users. Maybe make it so instead of .sav2 its NAME_OF_GAME_2.sav

Actually now that I think about it, most of what you're saying is a non-issue. If a user has a second Pokémon FireRed save for example, they must have done something on their own already to make that possible since there's no way mGBA would organically allow you to have two .sav files for the same game. So if they're smart enough to figure that out, they're smart enough to figure out that they need to rename (or choose via the UI) their .sav file to use it on a second mGBA instance.

**Comment by tony971** [comment]
Yea this isn't for the users that have a second save file. It's for the ones that get confused as to why player two is trying to load the same one.

**Comment by tony971** [comment]
https://moqups.com/tony971/QrOwORV0

Last one. This is fairly simple now. If a multiplayer window loads a ROM that's already running, then it should default to Cartridge 2. An OSD "Cartridge 2 loaded" should confirm this. Extrapolate this out to players 3 and 4 and you've got 4 "separate cartridges" playing together. The menu dialogue allows users to decide which cartridge to play with, allowing them to make significant progress through the game outside of multliplayer. Thoughts? @karasuhebi @endrift

**Comment by tony971** [comment]
Pinging @JMC47 because end-user relations is somewhat of a specialty of yours.

**Comment by JMC47** [comment]
I've actually corrupted saves and crashed games with the multiplayer stuff loading the same game.

Here's my suggestion:  If a save exists (let's pretend MegaMan(2).sav) then it uses that.  If no save exists, we have two options.

1: Automatically copy the first save and use that either temporarily or permanently.
2: Make a new save that saves with a (2) at the end of it.

All I know is that as someone who's ran into issues with my cousin trying to set thigns up, this would be really, really nice.

**Comment by tony971** [comment]
Personally, I'd rather use .sav2 files instead of (2).sav to maintain compatibility with V-GBA saves.

**Comment by JMC47** [comment]
I was just making up a random naming scheme, any works for me.

**Comment by tony971** [comment]
This should probably be done before finishing the Dolphin stuff because the flagship game for Dolphin connectivity is Four Swords Adventures. A lot of end users are going to fire up four instances of this game.

**Comment by JMC47** [comment]
There's no saving on the GBA involved for Four Swords Adventures.

**Comment by endrift** [comment]
Added in 667dffe515e1b36384bc06a9cf88281e99aec0eb, though it doesn't affect save states.

**Comment by profi200** [comment]
@endrift 
I saw this recently and wondered about the chosen extension. Many Windows users who don't know better probably don't have the option set to show file name extensions. So all they see is multiple identically named files.
I got this idea from DS flashcards. They use this format: "whatever.X.sav" where X is the savegame slot number. 0 (default) is the exception and has the ".X" omitted for compatibility.
